scripts/leapfrog/README.md

```diff
@@ -0,0 +1,136 @@
+* [OSA](https://github.com/openstack/openstack-ansible):  OpenStack Ansible
+* [OSA-OPS](https://github.com/openstack/openstack-ansible-ops):  OpenStack Operations
+* [Kilo](https://github.com/rcbops/rpc-openstack/tree/kilo): The RPCO release of OpenStack Kilo
+* [Newton-14.1](https://github.com/rcbops/rpc-openstack/tree/newton-14.1) The second RPCO release of OpenStack Newton. This version includes Leapfrog upgrade tools.
+
+## Executing a leapfrog upgrade
+
+The start point is a kilo deployment. It is useful to check the kilo deployment
+is healthy before starting the leapfrog process, for example by running tempest.
+
+```
+root@kilo:/opt/rpc-openstack/rpcd/playbooks# openstack-ansible ../../scripts/run_tempest.yml -t tempest_execute_tests -vv
+```
+
+If tempest raises issues, those should be fixed before continuing.
+
+The next step is to checkout the newton-14.1 branch of rpco. This will remove
+any uncommitted information from the rpco repo, so back that up if it is
+required.
+
+```
+root@kilo:/opt/rpc-openstack# git reset --hard
+git clean -df
+pushd openstack-ansible
+git reset --hard
+git clean -df
+popd
+git fetch origin
+git checkout newton-14.1
+git submodule update
+git status
+```
+
+It is crucial that the rpco repo and openstack-ansible submodule are both clean
+and at the appropriate SHAs for newton-14.1 at this stage. The [github
+newton-14.1 branch
+page](https://github.com/rcbops/rpc-openstack/tree/newton-14.1) shows both SHAs,
+the RPCO SHA is the latest commit, and the OSA SHA is shown as an annotation
+after @ on the openstack-ansible submodule.
+
+The next step is to execute the leapfrog upgrade script
+
+```
+root@kilo:/rpc-openstack# scripts/leapfrog/ubuntu14-leapfrog.sh
+```
+
+## Structure of the leapfrog process.
+<img align="right" src="docs/images/leapfrog_structure_diagram.png"/>
+The RPCO leapfrog scripts are a thin wrapper around OSA-OPS leapfrog tools. 
+
+For details please refer to the scripts themselves. Paths are omitted for
+brevity, scripts may not be in the root of the relevant repo.
+
+
+#### Pre Leap
+This step removes modifications to RPCO Kilo that aren't compatible with OSA
+Newton. Currently this only contains an Ansible 1.9 compatibility workaround.
+
+#### Prep
+This step executes pre-flight checks, and prompts the user for confirmation. It
+also ensures that the databases are backed up.
+
+#### Upgrade
+This step has a section for each major version between the source (Kilo) and
+target (Newton) versions. Each section includes:
+* Variable & Secrets Migration (OSA)
+* Fact Cleanup
+* Hostname Compatibility checks/modifications.
+
+#### Migrations
+The step runs the database migrations for each major upgrade in turn.
+For a RPC Kilo --> Newton upgrade this involves running the liberty,
+mitaka and newton migrations in sequence.
+
+#### Re-Deploy
+This step first runs the pre_re deploy script which handles RPC variable and
+secret migrations. Next the target version of OSA is deployed.
+
```

Comment:
with some additional detail on why it does that (to keep data plane access as long as possible during the upgrade).
Maybe each section at a high level should have an indicator of which bits are 'still up' and which bits are 'hard down'. Not how long, because we don't have the data yet, just which components are impacted during which phase.